keps/sig-api-machinery/1027-api-unions/README.md

```diff
@@ -183,133 +424,847 @@ for validation, but is "on-top" of this proposal.
 A new extension is created in the openapi to describe the behavior:
 `x-kubernetes-unions`.
 
-This is a list of unions that are part of this structure/object. Here is what
-each list item is made of:
-- `discriminator: <discriminator>` is set to the name of the discriminator
-  field, if present,
-- `fields-to-discriminateBy: {"<fieldName>": "<discriminateName>"}` is a map of
-  fields that belong to the union to their discriminated names. The
-  discriminatedValue will typically be set to the name of the Go variable.
+This is a list of unions that are part of this structure/object. Each item in
+the list represents a discriminator for the union, and for each member field,
+the discriminator value of that field and whether or not that field is optional.
 
 Conversion between OpenAPI v2 and OpenAPI v3 will preserve these fields.
 
-### Discriminator
+The following is an example of what the generated OpenAPI definition will look
+like for a given go type.
+
+```
+const (
+  FieldA Union1Type = "FieldA"
+  FieldB Union1Type = "FieldB"
+  FieldC Union1Type = "FieldC"
+  FieldD Union1Type = "FieldD"
+  FieldNone Union1Type = ""
+)
+const (
+  Alpha Union2Type = "ALPHA"
+  Beta Union2Type = "BETA"
+  Gamma Union2Type = "GAMMA"
+  Null Union2Type = "NULL"
+)
+
+// This will generate one union, with two fields and a discriminator.
+type Union struct {
+  // +unionDiscriminator
+  // +required
+  Union1 Union1Type `json:"union1"`
+
+  // +unionMember
+  // +unionDiscriminatedBy=Union1
+  // +optional
+  FieldA int `json:"fieldA"`
+  // +unionMember,optional
+  // +unionDiscriminatedBy=Union1
+  // +optional
+  FieldB int `json:"fieldB"`
+
+  // +unionDiscriminator
+  // +required
+  Union2 Union2Type `json:"union2"`
+
+  // +unionMember=ALPHA,
+  // +unionDiscriminatedBy=Union2
+  // +optional
+  Alpha int `json:"alpha"`
+  // +unionMember=BETA,optional
+  // +unionDiscriminatedBy=Union2
+  // +optional
+  Beta int `json:"beta"`
+}
+```
+
+turns into:
+```
+OpenAPIDefinition{
+  Schema: spec.Schema{
+    SchemaProps: spec.SchemaProps{
+      ... // schema props omitted
+    },
+    VendorExtensible: spec.VendorExtensible{
+      Extensions: spec.Extensions{
+        "x-kubernetes-unions": []interface{}{
+          map[string]interface{}{
```

Comment:
would this be more natural to place in the discriminator property, where we already have the enum?

```yaml
properties:
- name: union1
  enum: ["FieldA","FieldB","None"]
  x-kubernetes-union-members:
    FieldA:
      property: fieldA
      optional: true
    FieldB:
      property: fieldB
    None:
      property: ""
```

Reply:
that would keep us from having to repeat `discriminator` a lot, would let us drop a mostly unused level of structure (since multiple unions in a single struct are very unlikely), and would localize validation of the union member keys and the enum

validation could require an entry in x-kubernetes-union-members for every item in the enum, and could validate the existence of all referenced peer properties